File: live-location/src/main/java/com/singularity_code/live_location/util/other/AlienPortal.java
package com.singularity_code.live_location.util.other;

import android.content.Context;
import android.os.Build;
import androidx.core.app.NotificationCompat;
import com.google.android.gms.location.LocationResult;
import com.singularity_code.live_location.util.enums.NetworkMethod;
import com.singularity_code.live_location.util.pattern.*;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;

public class AlienPortal {

    public static AlienPortal newInstance() {
        return new AlienPortal();
    }

    // observable
    private String _status = "DEAD";

    public String getStatus() {
        return _status;
    }

    private String _errorMessage;

    public String getErrorMessage() {
        return _errorMessage;
    }

    private String _latitude;

    public String getLatitude() {
        return _latitude;
    }

    private String _altitude;

    public String getAltitude() {
        return _altitude;
    }

    private String _altitudeAccuracyMeter;
    public String getAltitudeAccuracyMeter() {
        return _altitudeAccuracyMeter;
    }
    private String _longitude;

    public String getLongitude() {
        return _longitude;
    }

    private String _accuracy;

    public String getAccuracy() {
        return _accuracy;
    }

    private String _updatedTime;

    public String getUpdatedTime() {
        return _updatedTime;
    }

    // gps config
    private long gpsSamplingRate = 10000L;

    // notification config
    private Context context;
    private String channelId;
    private String channelName;
    private String channelDescription;
    private String notificationTitle;
    private String notificationMessage;

    // network configuration
    private String apiOrSocketURL;
    private NetworkMethod networkMethod;
    private String messageDescriptor;
    private HashMap<String, String> headers = new HashMap<>();

    private LiveLocationServiceInteractor liveLocationServiceInteractor;

    public void setGPSSamplingRate(long samplingRate) {
        gpsSamplingRate = samplingRate;
    }

    public void setContext(Context context) {
        this.context = context;
    }

    public void setChannelId(String id) {
        channelId = id;
    }

    public void setChannelName(String name) {
        channelName = name;
    }

    public void setChannelDescription(String description) {
        channelDescription = description;
    }

    public void setupAPIorSocketURL(String url) {
        apiOrSocketURL = url;
    }

    public void setNotificationTitle(String title) {
        notificationTitle = title;
    }

    public void setNotificationMessage(String message) {
        notificationMessage = message;
    }

    // 0 -> RESTFUL
    // 1 -> Web Socket
    public void setupNetworkMethod(int methodEnumIndex) {
        networkMethod = NetworkMethod.values()[methodEnumIndex];
    }

    public void addHeader(String key, String value) {
        headers.put(key, value);
    }

    public void clearHeader() {
        headers.clear();
    }

    public void setMessageDescriptor(String descriptor) {
        messageDescriptor = descriptor;
    }

    public void start() {
        liveLocationServiceInteractor = new LiveLocationServiceInteractorAbs() {
            @Override
            public Context getContext() {
                return AlienPortal.this.context;
            }

            @NotNull
            @Override
            public GPSConfig getGpsConfig() {
                return new GPSConfig() {
                    @Override
                    public long getSamplingRate() {
                        return gpsSamplingRate;
                    }
                };
            }

            @Override
            public NotificationConfig getNotificationConfig() {
                return new NotificationConfig() {
                    @Override
                    public int getForegroundServiceID() {
                        return 1003;
                    }

                    @Override
                    public String getNotificationChannelID() {
                        return AlienPortal.this.channelId;
                    }

                    @Override
                    public String getNotificationChannelName() {
                        return AlienPortal.this.channelName;
                    }

                    @Override
                    public String getNotificationChannelDescription() {
                        return AlienPortal.this.channelDescription;
                    }

                    @Override
                    public int getNotificationPriority() {
                        return NotificationCompat.PRIORITY_DEFAULT;
                    }

                    @Override
                    public Integer getIconRes() {
                        return null;
                    }
                };
            }

            @Override
            public LiveLocationNetworkConfiguration getNetworkConfiguration() {
                return new LiveLocationNetworkConfiguration() {
                    @Override
                    public String getUrl() {
                        return AlienPortal.this.apiOrSocketURL;
                    }

                    @Override
                    public NetworkMethod getNetworkMethod() {
                        return AlienPortal.this.networkMethod;
                    }

                    @Override
                    public HashMap<String, String> getHeaders() {
                        return AlienPortal.this.headers;
                    }

                    @Override
                    public String getMessageDescriptor() {
                        return AlienPortal.this.messageDescriptor;
                    }
                };
            }

            @Override
            public void onServiceStatusChanged(LiveLocationServiceInteractor.ServiceStatus serviceStatus) {
                AlienPortal.this._status = serviceStatus.name();
            }

            @Override
            public void onError(String message) {
                AlienPortal.this._errorMessage = message;
            }

            @Override
            public void onReceiveUpdate(@NotNull LocationResult location) {
                if (location.getLastLocation() == null) return;

                AlienPortal.this._latitude = String.valueOf(location.getLastLocation().getLatitude());
                AlienPortal.this._longitude = String.valueOf(location.getLastLocation().getLongitude());
                AlienPortal.this._accuracy = String.valueOf(location.getLastLocation().getAccuracy());
                AlienPortal.this._updatedTime = String.valueOf(location.getLastLocation().getTime());

                if (location.getLastLocation().hasAltitude()) {
                    AlienPortal.this._altitude = String.valueOf(location.getLastLocation().getAltitude());

                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                        AlienPortal.this._altitudeAccuracyMeter = String.valueOf(location.getLastLocation().getVerticalAccuracyMeters());
                    }
                }
            }
        };

        liveLocationServiceInteractor.startService(
                AlienPortal.this.notificationTitle,
                AlienPortal.this.notificationMessage
        );
    }

    public void stop() {
        if (liveLocationServiceInteractor != null) {
            liveLocationServiceInteractor.stopService();
            liveLocationServiceInteractor = null;
        }
    }
}

File: live-location/src/main/java/com/singularity_code/live_location/util/other/AlienPortal.kt
package com.singularity_code.live_location.util.other

import android.content.Context
import androidx.core.app.NotificationCompat
import com.singularity_code.live_location.util.enums.NetworkMethod
import com.singularity_code.live_location.util.pattern.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

/**
 * In case you are using alien technologi such Xamarin/Unity, C++, or anything non java platform,
 * you can use this ALienPortal instance to interact with the LiveLocationService.
 */
class AlienPortal {

    // observable
    private var _status: String = "DEAD"
    fun getStatus() = _status

    private var _errorMessage: String? = null
    fun getErrorMessage() = _errorMessage

    private var _latitude: String? = null
    fun getLatitude(): String? = _latitude

    private var _longitude: String? = null
    fun getLongitude(): String? = _longitude

    private var _accuracy: String? = null
    fun getAccuracy(): String? = _accuracy

    private var _updatedTime: String? = null
    fun getUpdatedTime(): String? = _updatedTime

    // gps config
    private var gpsSamplingRate = 10000L

    // notification config
    private lateinit var context: Context
    private lateinit var channelId: String
    private lateinit var channelName: String
    private lateinit var channelDescription: String
    private lateinit var notificationTitle: String
    private lateinit var notificationMessage: String


    // network configuration
    private lateinit var apiOrSocketURL: String
    private lateinit var networkMethod: NetworkMethod
    private lateinit var messageDescriptor: String
    private var headers = hashMapOf<String, String>()

    private var liveLocationServiceInteractor: LiveLocationServiceInteractor? = null

    fun setGPSSamplingRate(samplingRate: Long) {
        gpsSamplingRate = samplingRate
    }

    fun setContext(context: Context) {
        this.context = context
    }

    fun setChannelId(id: String) {
        channelId = id
    }

    fun setChannelName(name: String) {
        channelName = name
    }

    fun setChannelDescription(description: String) {
        channelDescription = description
    }

    fun setupAPIorSocketURL(url: String) {
        apiOrSocketURL = url
    }

    fun setNotificationTitle(title: String) {
        notificationTitle = title
    }

    fun setNotificationMessage(message: String) {
        notificationMessage = message
    }

    // 0 -> RESTFUL
    // 1 -> Web Socket
    fun setupNetworkMethod(methodEnumIndex: Int) {
        networkMethod = NetworkMethod.values()[methodEnumIndex]
    }

    fun addHeader(key: String, value: String) {
        headers[key] = value
    }

    fun clearHeader() {
        headers.clear()
    }

    fun setMessageDescriptor(descriptor: String) {
        messageDescriptor = descriptor
    }

    fun start() {
        liveLocationServiceInteractor = object : LiveLocationServiceInteractorAbs() {
            override val context: Context = this@AlienPortal.context

            override val gpsConfig: GPSConfig =
                object : GPSConfig {
                    override val samplingRate: Long = 10000
                }

            override val notificationConfig: NotificationConfig =
                object : NotificationConfig {
                    override val foregroundServiceID: Int = 1003
                    override val notificationChannelID: String = this@AlienPortal.channelId
                    override val notificationChannelName: String = this@AlienPortal.channelName
                    override val notificationChannelDescription: String =
                        this@AlienPortal.channelDescription
                    override val notificationPriority: Int = NotificationCompat.PRIORITY_DEFAULT
                    override val iconRes: Int? = null
                }

            override val networkConfiguration: LiveLocationNetworkConfiguration =
                object : LiveLocationNetworkConfiguration {
                    override val url: String = this@AlienPortal.apiOrSocketURL
                    override val networkMethod: NetworkMethod = this@AlienPortal.networkMethod
                    override val headers: HashMap<String, String> = this@AlienPortal.headers
                    override val messageDescriptor: String by lazy {
                        this@AlienPortal.messageDescriptor
                    }
                }

            private val coroutine by lazy {
                CoroutineScope(Dispatchers.IO + SupervisorJob())
            }

            override fun onServiceStatusChanged(
                serviceStatus: LiveLocationServiceInteractor.ServiceStatus
            ) {
                coroutine.launch {
                    this@AlienPortal._status = serviceStatus.name
                }
            }

            override fun onError(
                message: String?
            ) {
                coroutine.launch {
                    this@AlienPortal._errorMessage = message
                }
            }

            override fun onReceiveUpdate(
                latitude: Double,
                longitude: Double,
                accuracy: Float,
                updateTime: Long
            ) {
                coroutine.launch {
                    this@AlienPortal._latitude = latitude.toString()
                    this@AlienPortal._longitude = longitude.toString()
                    this@AlienPortal._accuracy = accuracy.toString()
                    this@AlienPortal._updatedTime = updateTime.toString()
                }
            }

        }

        liveLocationServiceInteractor?.startService(
            this@AlienPortal.notificationTitle,
            this@AlienPortal.notificationMessage
        )
    }

    fun stop() {
        liveLocationServiceInteractor?.stopService()
        liveLocationServiceInteractor = null
    }
}